docs/design/sql/16-partition-aware-client.md

```diff
@@ -0,0 +1,48 @@
+# Partition-aware client
+
+If we want SQL to be the primary interface to Hazelcast, the basic map get/put
+operations using SQL have to be equally fast as their native counterparts. We’ve
+implemented shortcut evaluation for the queries where the WHERE clause contains
+`__key=?` - we don’t start a job in this case, but directly execute
+`IMap.get()`. It’s also implemented for INSERT/SINK, UPDATE and DELETE
+statements.
+
+However, the operation is still much slower than the `IMap` counterparts. The
+main reason is that the client sending a query picks a random member to execute
+the query, unaware of the partition owner.
+
+The original idea to fix this was to use a kind of client-side plan cache. Along
+with the query result, the cluster will send a plan to the client telling it
+that the next time it sees that query, instead of sending it as a query it
+should execute IMap.get directly. There are two issues with this approach:
+
+1. cache invalidation: the member will have to keep track what plans were cached
+   by which client, and will need to invalidate those 
+
+2. client-side complexity: each type of operation will have to be implemented by
+   each client. It’s not just simple `get`, but `get` with projections, additional
+   filters, and also `put` for INSERT, and entry processor for UPDATE, which might
+   not be possible to implement on non-java client.
```

Comment:
What will we do if a client does not have the required feature? I guess the client will be able to use old school SQL query. It will be slower than the optimized approach, but at least it will work.